text_extensions_for_pandas/array/arrow_conversion.py

```diff
@@ -0,0 +1,327 @@
+
+
+class ArrowTokenSpanType(pa.PyExtensionType):
+    """
+    PyArrow extension type definition for conversions to/from TokenSpan columns
+    """
+
+    TARGET_TEXT_KEY = ArrowCharSpanType.TARGET_TEXT_KEY
+    BEGINS_NAME = "token_begins"
+    ENDS_NAME = "token_ends"
+
+    def __init__(self, index_dtype, target_text):
+        """
+        Create an instance of a TokenSpan data type with given index type and
+        target text that will be stored in Field metadata.
+
+        :param index_dtype:
+        :param target_text:
+        """
+        assert pa.types.is_integer(index_dtype)
+
+        # Store target text as field metadata
+        metadata = {self.TARGET_TEXT_KEY: target_text}
+
+        fields = [
+            pa.field(self.BEGINS_NAME, index_dtype),
+            pa.field(self.ENDS_NAME, index_dtype),
+            pa.field(ArrowCharSpanType.BEGINS_NAME, index_dtype, metadata=metadata),
+            pa.field(ArrowCharSpanType.ENDS_NAME, index_dtype)
+        ]
+
+        pa.PyExtensionType.__init__(self, pa.struct(fields))
+
+    def __reduce__(self):
+        index_dtype = self.storage_type[self.BEGINS_NAME].type
+        metadata = self.storage_type[ArrowCharSpanType.BEGINS_NAME].metadata
+        return ArrowTokenSpanType, (index_dtype, metadata)
+
+
+def char_span_to_arrow(char_span):
+    """
+    Convert a CharSpanArray to a pyarrow.ExtensionArray with a type
+    of ArrowCharSpanType and struct as the storage type. The resulting
+    extension array can be serialized and transferred with standard
+    Arrow protocols.
+
+    :param char_span: A CharSpanArray to be converted
+    :return: pyarrow.ExtensionArray containing CharSpan data
+    """
+    # Create array for begins, ends
+    begins_array = pa.array(char_span.begin)
+    ends_array = pa.array(char_span.end)
+
+    typ = ArrowCharSpanType(begins_array.type, char_span.target_text)
+    fields = list(typ.storage_type)
+
+    storage = pa.StructArray.from_arrays([begins_array, ends_array], fields=fields)
+
+    return pa.ExtensionArray.from_storage(typ, storage)
+
+
+def arrow_to_char_span(extension_array):
+    """
+    Convert a pyarrow.ExtensionArray with type ArrowCharSpanType to
+    a CharSpanArray.
+
+    :param extension_array: pyarrow.ExtensionArray with type ArrowCharSpanType
+    :return: CharSpanArray
+    """
+    if isinstance(extension_array, pa.ChunkedArray):
+        if extension_array.num_chunks > 1:
+            raise ValueError("Only pyarrow.Array with a single chunk is supported")
+        extension_array = extension_array.chunk(0)
+
+    assert pa.types.is_struct(extension_array.storage.type)
+
+    # Get target text from the begins field metadata and decode string
+    metadata = extension_array.storage.type[ArrowCharSpanType.BEGINS_NAME].metadata
+    target_text = metadata[ArrowCharSpanType.TARGET_TEXT_KEY]
+    if isinstance(target_text, bytes):
+        target_text = target_text.decode()
```

Comment:
Should this line specify a specific encoding?

Reply:
I don't think it's necessary, pyarrow does the same I believe